aa-sdk/core/src/middleware/erc7677middleware.ts

```diff
@@ -159,21 +185,143 @@ export function erc7677Middleware<
 
   const paymasterAndData: ClientMiddlewareFn = async (
     uo,
-    { client, account, feeOptions, overrides }
+    { client, account, feeOptions, overrides, policyToken }
   ) => {
     const userOp = deepHexlify(await resolveProperties(uo));
-    const context =
+
+    let context =
       (typeof params?.context === "function"
         ? await params?.context(userOp, { overrides, feeOptions })
         : params?.context) ?? {};
+    const entrypoint = account.getEntryPoint();
 
     if (!client.chain) {
       throw new ChainNotFoundError();
     }
 
+    let max_token: BigInt | undefined = undefined;
+    if (policyToken !== undefined) {
+      let total_gas = BigInt(0);
+      total_gas += uo.callGasLimit
+        ? BigInt(uo.callGasLimit.toLocaleString())
+        : BigInt(0);
+
+      total_gas += uo.preVerificationGas
+        ? BigInt(uo.preVerificationGas.toLocaleString())
+        : BigInt(0);
+
+      total_gas += uo.verificationGasLimit
+        ? BigInt(uo.verificationGasLimit.toLocaleString())
+        : BigInt(0);
+
+      let paymasterAddress = "0x";
+
+      if (entrypoint.version === "0.7.0") {
+        const uo_07 = uo as UserOperationStruct<"0.7.0">;
```

Comment:
Just to reiterate: `uo` is not a `UserOperationStruct<"0.7.0">`, it is a `PromiseOrValue<UserOperationStruct<"0.7.0">>`. You should read from `userOp` instead.